### ISSUE TITLE
[feature] merge hash variable

### ISSUE THREAD

**Issue by andreineculau** [tomasbjerre/pull-request-notifier-for-bitbucket#63]
currently it is possible to get the hash of the source and the target branch, meaning that if you want to "recreate" the result of the pull-request, you have to merge the source into the target (or rebase the source on top of the target)

BUT that process has already been done by Stash in `refs/pull-requests/<ID>/merge`

Would it be possible to export that into a variable e.g. `PULL_REQUEST_MERGE_HASH`? That way one can use the merge hash to communicate (logs, other systems) about the exact same commit rather than the tuple `{FROM_HASH, TO_HASH}`

**Comment by itay** [comment]
@andreineculau unfortunately it's not so easy. It turns out the behavior of the `refs/pull-requests/...` tree is very subtle. Namely, that ref is not generated or updated unless you visit the UI first - i.e. it is done on-demand.

That means that at the time of the PR creation, there is no "merge hash" for which you can be notified. Additionally, you don't want to cause the server to generate these always, as it is very expensive (per Atlassian comments), and you're better off doing it yourself on the CI system (e.g. Jenkins), to reduce the load on the Stash server.

**Comment by andreineculau** [comment]
@itay

> @andreineculau unfortunately it's not so easy. It turns out the behavior of the refs/pull-requests/... tree is very subtle. Namely, that ref is not generated or updated unless you visit the UI first - i.e. it is done on-demand.

hmm, I cannot reproduce that behaviour on Stash 3.11.1 (tested via `git ls-remote`)

but despite this being an unexpected behaviour (the expected one being that every push to the source/target branch would update `refs/pull-requests/ID/from` and `refs/pull-requests/ID/merge`) I don't see that as relevant for this ticket since we are talking about a UI plugin for stash. I could easily buy why this assumption might be wrong:

1. I don't know (didn't check the code) how this addon listens on RESCOPED* events for instance
2. if you use the comment trigger, then a comment can be added via other interfaces than the UI (e.g. REST), thus once again the hash wouldn't be updated

> That means that at the time of the PR creation, there is no "merge hash" for which you can be notified.

I haven't tested that, but that's also ok IMHO, again despite the unexpected behaviour.

> Additionally, you don't want to cause the server to generate these always, as it is very expensive (per Atlassian comments), and you're better off doing it yourself on the CI system (e.g. Jenkins), to reduce the load on the Stash server.

Well, I leave aside that I don't buy that for 9/10 cases, no matter the size of repo/team, but thats also not relevant.

*In a nutshell, implementation details aside, that hash, whenever available and no matter how out-of-sync, could be exposed by this addon.*

**Comment by andreineculau** [comment]
@itay now you made me curious :) where did you learn all this info, and specifically the "expensive per Atlassian comments" part? Thanks!

**Comment by itay** [comment]
@andreineculau you can see some of this information here:
https://answers.atlassian.com/questions/239988/change-pull-request-refs-after-commit-instead-of-after-approval-or-workaround

and here:

https://github.com/Nerdwin15/stash-jenkins-postreceive-webhook/issues/83

Overall, we've noticed the same thing. We are also in the 1/10 situation you described - we have a very large Stash installation (using Stash DC), with hundreds of PRs every day. We definitely see slowdowns when a lot of "simulated merges" are happening on that ref (we track this on the machines).

**Comment by tomasbjerre** [comment]
Ok to close this one?

**Comment by tomasbjerre** [comment]
Open again if we should do something here.
